### DISCUSSION TITLE
Why not support MLIR instead of LLVM?

### DISCUSSION THREAD

**Discussion by takatoyamada** [carbon-language/carbon-lang#1913] [General]
Recently MLIR is getting well-known as a more flexible and high-performance compiler backend rather than using pure LLVM.
There are already some projects supporting MLIR even to compile C++ such as [Polygeist](https://github.com/llvm/Polygeist), [clangir](https://github.com/facebookincubator/clangir), [cil(deprecated?)](https://github.com/compiler-tree-technologies/cil) since it can represent high-level abstraction like structured control flow, affine, vector and etc... (I understand C++ frontend cannot easily express them).
I'm wondering if there is any plan to use MLIR in Carbon language backend.

**Comment by niekbouman** [comment]
@chandlerc 
I was having the same question, especially based on the ["MLIR Primer: A Compiler Infrastructure for the End of Moore’s Law"](https://pdfs.semanticscholar.org/d43c/f123628324291328ddd0b3966dfa6f338e25.pdf)  talk (2019) by Chris Lattner.
Is it perhaps the case that the goals/requirements of the Carbon toolchain are too far away from MLIR's goals/implementation ?

![image](https://github.com/carbon-language/carbon-lang/assets/2826915/f86ee7a2-ed89-4459-8ad3-ef08aa960ae9)

**Reply by chandlerc** [reply]
I added a top level response to the original question.

In general, I do think that there is a fundamental tension between a goal of maximum flexibility and maximum efficiency. Sometimes, you can achieve both (although there is always a cost, usually complexity), but often there isn't a good way to *fully* achieve both, and at least initially, our priority is both efficiency and simplicity, which motivate us to look in a different direction.

**Comment by chandlerc** [comment]
No current plans -- my impression at least is that the overall design of MLIR is more focused on flexibility and progressive lowering. This is really impactful for a lot of high-level domain specific languages and problems (from numerical stuff to image manipulation pipelines, and obviously ML workloads). But typically, this kind of flexibility comes with a tradeoff of efficiency and large-scale. At least initially, we want to establish a baseline that deeply prioritizes large-scale and efficient representation of the program in the toolchain, and a more minimal and purpose-built set of data structures seems more effective for that.

Of course, we may end up with a hybrid where some kernels are expressed in a higher level way and lower through MLIR or similar infrastructure, but the rest of the application software uses a lowering path designed for efficient scaling. But right now, or focus is on the baseline and getting enough information to decide on whether this experiment is worthwhile.

**Comment by zygoloid** [comment]
I [gave a lightning talk](https://youtu.be/vIWT4RhUcyw) at EuroLLVM earlier this year ([slides](https://llvm.org/devmtg/2024-04/slides/LightningTalks/Smith-Carbons-high-level-semanticIR.pdf)) that also touches on this question.

**Comment by rdong8** [comment]
I think it's worth revisiting this discussion now that Mojo has hit 1.0.0. They go as far as using MLIR for the AST itself, among a bunch of other interesting techniques https://smolnero.com/posts/first-principle-mojo-kgen-mlir-and-what-a-compiler-chooses-to-remember
